platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java

```diff
@@ -16,54 +16,58 @@
 
 package org.gradle.api.internal.tasks.compile;
 
-import org.gradle.api.JavaVersion;
 import org.gradle.api.tasks.compile.CompileOptions;
 import org.gradle.internal.Factory;
 import org.gradle.internal.jvm.Jvm;
 import org.gradle.jvm.toolchain.JavaInstallationMetadata;
 import org.gradle.jvm.toolchain.internal.JavaExecutableUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import javax.annotation.Nullable;
 import java.io.File;
 
 public abstract class AbstractJavaCompileSpecFactory<T extends JavaCompileSpec> implements Factory<T> {
-    private final CompileOptions compileOptions;
+    private final static Logger LOGGER = LoggerFactory.getLogger(AbstractJavaCompileSpecFactory.class);
 
+    private final CompileOptions compileOptions;
     private final JavaInstallationMetadata toolchain;
 
-    public AbstractJavaCompileSpecFactory(CompileOptions compileOptions, @Nullable JavaInstallationMetadata toolchain) {
+    public AbstractJavaCompileSpecFactory(CompileOptions compileOptions, JavaInstallationMetadata toolchain) {
         this.compileOptions = compileOptions;
         this.toolchain = toolchain;
     }
 
     @Override
     public T create() {
-        if (toolchain != null) {
-            return chooseSpecForToolchain();
-        }
-
-        if (compileOptions.isFork()) {
-            return chooseSpecFromCompileOptions(Jvm.current().getJavaHome());
-        }
-
-        return getDefaultSpec();
-    }
-
-    private T chooseSpecForToolchain() {
         File toolchainJavaHome = toolchain.getInstallationPath().getAsFile();
         if (!toolchain.getLanguageVersion().canCompileOrRun(8)) {
+            LOGGER.info("Falling back to command line compiler: requested toolchain is below JDK 1.7");
             return getCommandLineSpec(Jvm.forHome(toolchainJavaHome).getJavacExecutable());
         }
 
         if (compileOptions.isFork()) {
-            return chooseSpecFromCompileOptions(toolchainJavaHome);
+            File forkJavaHome = compileOptions.getForkOptions().getJavaHome();
+            if (forkJavaHome != null) {
+                LOGGER.info("Falling back to command line compiler: forking with Java Home set");
+                return getCommandLineSpec(Jvm.forHome(forkJavaHome).getJavacExecutable());
+            }
+
+            String forkExecutable = compileOptions.getForkOptions().getExecutable();
+            if (forkExecutable != null) {
+                LOGGER.info("Falling back to command line compiler: forking with javac executable set");
+                return getCommandLineSpec(JavaExecutableUtils.resolveExecutable(forkExecutable));
+            }
         }
 
-        if (!toolchain.isCurrentJvm()) {
-            return getForkingSpec(toolchainJavaHome, toolchain.getLanguageVersion().asInt());
+        if (toolchain.isCurrentJvm() && JdkJavaCompiler.canBeUsed()) {
```

Comment:
Ah, I got this slightly wrong.  This now allows things to run in process when compileOptions.fork = true and the toolchain is the current JVM.

I think this needs to be made an `else if` with the block above.

Reply:
Yeah, haven't seen this and realized myself :D
Had a deeper look at the diff, and implemented so the `isFork` branch always returns a spec.